### ISSUE TITLE
Tests with ncat (or with a server started in the background from a command substitution)

### ISSUE THREAD

**Issue by x-yuri** [bats-core/bats-core#1035]
This is unlikely a bug, but a problem I ran into while writing tests, and for a while couldn't find a solution. Although solution seems to be found (don't start a server from a command substitution), I'd like to understand it better. And since I have little to no knowledge of how `bats` works internally (at least the part I ran into), I could use some help here.

I tried to start `ncat` in the background... the idea was to create a helper, but basically:

```bats
@test whatever {
    pid=`ncat -l localhost 1024 & echo $!`
    ...
}
```

And it hang at the `pid=...` line.

Now I understand that `bash` here creates a pipe one end of which it hands over to the subshell (`ncat ... $!`) as `stdout`. `ncat` inherits this end of the pipe, doesn't close it, which blocks `bash` ([here][a]). At least that's my understanding.

A simple way to reproduce this is:

```
$ bash -c 'pid=`ncat -l localhost 1024 & echo $!`; kill "$pid"'
```

And a simple way to overcome this is:

```
$ bash -c 'pid=`ncat -l localhost 1024 >/dev/null & echo $!`; kill "$pid"'
```

But with `bats` it doesn't end here. It doesn't work even with the following test:

```bats
@test whatever {
    pid=`ncat -l localhost 1024 >dev/null 2>&1 </dev/null & echo $!`
    ...
}
```

Although the `pid=...` line finishes, it hangs somewhere down the road. What I was able to figure out, if I have a dummy test (`@test whatever { true }`), and put the following line:

```
pid=`ncat -l localhost 1024 >dev/null 2>&1 </dev/null & echo $!`
```

before the [following line][b], it finishes. But if I put it after the line, it hangs.

I can see that `ncat` shares a pipe (fd 4) with another process (fd 0) with args `bash ~/.bats/libexec/bats-core/bats a.bats` (there's a parent with the same args, so this must be some subshell of the script), but I don't understand what and why blocks. Killing `ncat` unblocks it.

The goal of course is to understand why the `pid=...` line (with `>/dev/null 2>&1 </dev/null`) doesn't work in a test.

Could you possibly explain what's going on or give some hints? Or what can I do to investigate further?

[a]: https://git.savannah.gnu.org/cgit/bash.git/tree/subst.c?id=6794b5478f660256a1023712b5fc169196ed0a22#n6818
[b]: https://github.com/bats-core/bats-core/blob/b640ec3cf2c7c9cfc9e6351479261186f76eeec8/libexec/bats-core/bats-exec-suite#L183

**Comment by profesorasix** [comment]
Hi, similiar problem here following the tutorial 
```bash
PORT=$(project.sh start-echo-server 2>&1 >/dev/null)
```
hangs until I kill ncat.

**Comment by x-yuri** [comment]
@profesorasix In your case... I don't think you want to keep `ncat` running after tests have finished:

```bats
setup() {
    [ "$BATS_LIB_PATH" = /usr/lib/bats ] && BATS_LIB_PATH=$HOME/.bats/lib:$BATS_LIB_PATH
    bats_load_library bats-support
    bats_load_library bats-assert
}

@test "server is reachable" {
    ncat -lk localhost 2000 -c 'echo test' &
    run ncat --recv-only localhost 2000
    assert_output test
}

teardown() {
    kill $!
}
```

More info on the test environment can be found [here](https://gist.github.com/x-yuri/e5b0baabc76e5e4a00d3e1fb40ebe256)

What I ran into... I was trying to create a helper that can be used in other test files and outputs the pid of the server:

```bats
setup() {
    [ "$BATS_LIB_PATH" = /usr/lib/bats ] && BATS_LIB_PATH=$HOME/.bats/lib:$BATS_LIB_PATH
    bats_load_library bats-support
    bats_load_library bats-assert
}

run_ncat_server() {
    ncat -lk localhost 2000 -c 'echo test' >/dev/null &
    # wait4ports tcp://localhost:2000 >/dev/null
    echo $!
}

@test "server is reachable" {
    pid=`run_ncat_server`
    run ncat --recv-only localhost 2000
    assert_output test
}

teardown() {
    kill -- "$pid"
}
```

But actually I just tried and it works. Even with the `wait4ports` line. I guess my issue was that I ran `wait4ports` without `> /dev/null`. But then, why does it hang with `wait4ports` without `>/dev/null`? Okay, `kill` will later fail, but why does it hang? `wait4ports` is not a server that keeps the command substitution's pipe open.

**Comment by glencornell** [comment]
I don't think this is a bug with BATS.  Rather, it's just a side effect of how BATS handles background jobs.  Consider using ncat's `--idle-timeout` option (or the `timeout` command) to force the background process to terminate after some reasonable time:

`
function find_free_tcp_port() {
   # open a socket and bind to an ephemeral port then close it.  It's not likely to be used again soon.
    python3 <<EOF                                                                                                                                                                                         
import socket                                                                                                                                                                                             
with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:                                                                                                                                              
    s.bind(('', 0))                                                                                                                                                                                       
    print(s.getsockname()[1])                                                                                                                                                                             
EOF                                                                                                                                                                                                       
}

setup() {
  FREE_TCP_PORT=$(find_free_tcp_port)
  OUTPUT_FILE=$(mktemp)
}

teardown() {
    rm -f "${OUTPUT_FILE}"
}

@test "tcp-ipv4-client" {
    local -r expected="Hello"
    ncat -vvv --recv-only --idle-timeout 10 -l4 -o ${OUTPUT_FILE} localhost ${FREE_TCP_PORT} &
    local -r server_pid=$!
    wait4ports tcp://localhost:${FREE_TCP_PORT}
    echo "${expected}" | ncat -vvv --send-only --idle-timeout 5 -4 localhost ${FREE_TCP_PORT}
    if kill -0 ${server_pid} 2>/dev/null; then
        kill ${server_pid} || true
    fi
    grep -q "${expected}" ${OUTPUT_FILE}
}
`
